File: services/spring-boot/product/src/main/java/product/core/cart/CartController.java
package product.core.cart;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.DeleteMapping;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PathVariable;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.PutMapping;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RequestParam;
import org.springframework.web.bind.annotation.RestController;

import product.core.cart.CartDto.AddProductCartDto;
import product.core.cart.CartDto.CreateCartDto;
import product.core.cart.CartDto.GetCartDto;
import product.core.cart.CartDto.ProductCartDto;
import product.core.cart.CartDto.UpdateCartDto;

@RestController
@RequestMapping("cart")
public class CartController {
    @Autowired
    private CartService service;

    @PostMapping()
    public ResponseEntity<CreateCartDto> createCart(@RequestBody() ProductCartDto product) {
        return ResponseEntity.ok(this.service.create(product));
    }

    @GetMapping("{id}")
    public ResponseEntity<GetCartDto> getCart(@PathVariable("id") String id) {
        return ResponseEntity.ok(this.service.get(id));
    }

    @PutMapping()
    public void updateCart(@RequestBody UpdateCartDto cart) {
        this.service.updateCart(cart.getId(), cart.getProducts());
    }

    @PostMapping("product")
    public void addToCart(@RequestBody AddProductCartDto productDto) {
        this.service.addToCart(productDto.getCartId(), productDto.getProduct());
    }

    @DeleteMapping("product/{id}")
    public void deleteFromCart(@PathVariable("id") int productId, @RequestParam("cart_id") String cartId) {
        this.service.deleteFromCart(productId, cartId);
    }
}

File: services/api-gateway/src/modules/cart/cart.controller.ts
import { Body, Controller, Delete, Get, Post, Put, Req } from '@nestjs/common'
import { Request, Response } from 'express'
import { CartService } from './cart.service'

@Controller('cart')
export class CartController {
  constructor(private readonly service: CartService) {}

  @Get(':id')
  public getCart(@Req() req: Request) {
    return this.service.get(req.url)
  }

  @Post()
  public createCart(@Req() req: Request, @Body() body) {
    return this.service.post(req.url, body)
  }

  @Post('product')
  public addToCart(@Req() req: Request, @Body() body) {
    return this.service.post(req.url, body)
  }

  @Delete('product/:id')
  public deleteFromCart(@Req() req: Request) {
    return this.service.delete(req.url)
  }

  @Post('test')
  public test(@Body() body) {
    return body
  }

  @Put()
  public updateCart(@Req() req: Request, @Body() body) {
    return this.service.put(req.url, body)
  }
}
